# External Market Salary Data

I looked into the external market to see if the company was loosing people due to the fact that salaries were not competitive

In [100]:
# Import necessary libraries

import requests
from bs4 import BeautifulSoup
# Collect first page of artists’ list
page = requests.get('http://www.usawage.com/high-pay/jobs.php')
#The salary and the employment information in America is computed from data published by U.S. Bureau of Labor Statistics (BLS) in April 2018 [1].
#I found a page in the internet and I scrapped just the table
#I made sure that the page was available
page

<Response [200]>

In [23]:
# Create a BeautifulSoup object
soup = BeautifulSoup(page.text, 'html.parser')


In [24]:
stat_table=soup.find_all('table')
stat_table
#I look at the content to make sure it coincides with what I want to scrape
print(stat_table[0])


<table class="table table-condensed2">
<tbody>
<tr>
<th><span>Rank</span></th>
<th class="">Job Description</th>
<th class="ar"> Average <br/> Salary </th>
<th class="ar" title="Number of employed"> # of <br/> Employment </th>
<!-- <th title="Major occupation group">Category<i class="badge_i">i</i></th> -->
</tr>
<tr><td>1</td>
<td> <a href="/usa/family_and_general_practitioners-salary.php" title="Detail salary information of family and general practitioners in America">Family and General Practitioners</a></td>
<td><div class=""> $198,740 </div></td>
<td class="ar"> 126,440</td></tr>
<tr><td>2</td>
<td> <a href="/usa/internists_general-salary.php" title="Detail salary information of general internists in America">General Internists</a></td>
<td><div class=""> $192,930 </div></td>
<td class="ar"> 42,280</td></tr>
<tr><td>3</td>
<td> <a href="/usa/dentists_all_other_specialists-salary.php" title="Detail salary information of other specialists dentists in America">Other Specialists Dentis

In [26]:
#to make sure I only grab one table
len(stat_table)

1

In [27]:
#making sure i only grab the only result for table
stat_table=stat_table[0]

In [28]:
type(stat_table)

bs4.element.Tag

In [121]:
# I tried to pull out the data but didnot manage, to
#properly allocate the data.  So i will need a 'for loop'

# df = pd.DataFrame(data) #Create a DataFrame from the data
# #print(len(df)) #See how long the dataset is (number of rows)
# df.columns = ['Rank', 'Job', 'Salary', 'NunberEd']
# df.head()

Rank   Job Salary  NunberEd
0  None  None      1   126,440
1  None  None      2    42,280
2  None  None      3     4,590
3  None  None      4       430
4  None  None      5   210,160

In [29]:
#I try to populate different lists
#I am going through each row (tr) and look for different 
#cell elements (td) to upload the different lists

listheaders=[]
listRank=[]
listJob=[]
listSalary=[]
listNumberEd=[]
all_rows = stat_table.find_all('tr')
#count= counts rows
count=0

for r in all_rows:
    ranking = ""
    job = ""
    salary=""
    ed = ""
    col = 0
    if count > 0:
        all_td= r.find_all("td")
        for d in all_td:
            if col==0:
                ranking=d.get_text()
                listRank.append(ranking)
            
            elif col==1:
                job=d.get_text()
                listJob.append(job)
            
            elif col==2:
                salary=d.get_text()
                listSalary.append(salary)
            elif col==3:
                ed=d.get_text()
                listNumberEd.append(ed) 
            else:
                print("nada")
                
            col=col+1
        
    else:
        
        allheaders = r.find_all("th")
        col = 0
        for x in allheaders:
            listheaders.append(x.get_text())
        
    
    count=count+1
    


In [23]:
print(listheaders)

['Rank', 'Job Description', ' Average  Salary ', ' # of  Employment ']


In [9]:
import pandas as pd


#listRank.pop()#run only once. There was an extra (td) for rank,so i had to delete it
print(len(listheaders))
print(len(listRank))
print(len(listJob))
print(len(listSalary))
print(len(listNumberEd))
d={'Rank':listRank,'Job Description':listJob, 'Average  Salary':listSalary, '# of  Employment':listNumberEd}
#with this information I created a data frame
df = pd.DataFrame(data=d)
df

4
200
200
200
200


Rank                                    Job Description Average  Salary  \
0      1                   Family and General Practitioners       $198,740    
1      2                                 General Internists       $192,930    
2      3                         Other Specialists Dentists       $190,840    
3      4                                    Prosthodontists       $185,150    
4      5                                   Chief Executives       $183,270    
5      6                              General Pediatricians       $172,650    
6      7                                 Nurse Anesthetists       $165,120    
7      8                                           Dentists       $158,120    
8      9                                   General Dentists       $151,440    
9     10          Computer and Information Systems Managers       $139,220    
10    11             Architectural and Engineering Managers       $137,720    
11    12     Airline Pilots, Copilots, and Flight Engineers       $137,330    
12    13         Judges, Magistrate Judges, and Magistrates       $133,840    
13    14                                Petroleum Engineers       $132,280    
14    15                                 Marketing Managers       $132,230    
15    16                                        Podiatrists       $127,740    
16    17                       Marketing and Sales Managers       $125,290    
17    18                                 Financial Managers       $125,080    
18    19                            Air Traffic Controllers       $124,540    
19    20                                        Pharmacists       $124,170    
20    21                                     Sales Managers       $121,060    
21    22                                            Lawyers       $119,250    
22    23                 Compensation and Benefits Managers       $119,120    
23    24                          Natural Sciences Managers       $118,970    
24    25                                         Physicists       $118,830    
25    26                         Astronomers and Physicists       $117,220    
26    27                    Lawyers and Judicial Law Clerks       $117,020    
27    28                                Purchasing Managers       $115,760    
28    29                        Computer Hardware Engineers       $115,120    
29    30                               Political Scientists       $115,110    
..   ...                                                ...             ...   
170  171                                  Dental Hygienists        $74,070    
171  172         Environmental Scientists and Geoscientists        $73,860    
172  173   Other Health Diagnosing and Treating Practiti...        $73,830    
173  174           Postsecondary Math and Computer Teachers        $73,780    
174  175                  Postsecondary Psychology Teachers        $73,770    
175  176                                     Ship Engineers        $73,110    
176  177                   Postsecondary Sociology Teachers        $73,080    
177  178                                    Gaming Managers        $72,930    
178  179                     Emergency Management Directors        $72,760    
179  180                     Postsecondary History Teachers        $72,690    
180  181   Postsecondary Area, Ethnic, and Cultural Stud...        $72,230    
181  182                          Transportation Inspectors        $72,140    
182  183         Occupational Health and Safety Specialists        $71,780    
183  184   First-Line Supervisors of Non-Retail Sales Wo...        $71,650    
184  185              Postsecondary Home Economics Teachers        $71,630    
185  186                            Producers and Directors        $71,620    
186  187                        Urban and Regional Planners        $71,490    
187  188                    Diagnostic Medical Sonographers        $71,410    
188  189                                    Credit Ana

In [24]:
#converting the '# of Employment' into a number
listNumberEd = [int(''.join(s[1:].split(','))) for s in listNumberEd]
print(listNumberEd)

[126440, 42280, 4590, 430, 210160, 28990, 42620, 125300, 110400, 365690, 179990, 84070, 28440, 32010, 218970, 9670, 590380, 569380, 22790, 309330, 371410, 628370, 15520, 56210, 16710, 18720, 643630, 70430, 66770, 6320, 27920, 65760, 122570, 67300, 37240, 136310, 34500, 394590, 28100, 49030, 16700, 430120, 16900, 109220, 157830, 17320, 166280, 2730, 19550, 134930, 33500, 849230, 19210, 2020, 6530, 171520, 2212200, 7470, 70820, 37500, 346980, 12770, 318300, 194610, 1617400, 12880, 131500, 32380, 44780, 105250, 183370, 14480, 27200, 250280, 6150, 270100, 6010, 38110, 118680, 142160, 8940, 263480, 27380, 10960, 200920, 69370, 28520, 687210, 477750, 315830, 26130, 250190, 581960, 20100, 104860, 2000, 10730, 13710, 920, 646570, 113690, 225420, 52640, 10800, 291660, 265520, 291290, 298910, 294110, 36540, 126050, 584940, 12110, 659200, 247690, 111690, 52580, 36190, 5770, 16200, 106050, 51530, 375040, 118560, 17250, 52290, 6850, 84340, 147140, 7280, 6350, 105350, 62710, 2770, 19030, 24490, 3549

In [10]:
len( listNumberEd)==len( listRank)==len(listJob)==len(listSalary)

True

In [17]:
#making sure list of salary is also an integer

print(listSalary)

[198740, 192930, 190840, 185150, 183270, 172650, 165120, 158120, 151440, 139220, 137720, 137330, 133840, 132280, 132230, 127740, 125290, 125080, 124540, 124170, 121060, 119250, 119120, 118970, 118830, 117220, 117020, 115760, 115120, 115110, 114520, 113030, 111930, 111280, 110300, 110120, 108250, 107600, 106130, 105840, 105810, 105610, 104910, 104860, 104650, 103990, 103880, 103010, 102490, 102180, 102160, 101790, 101560, 100590, 100590, 100580, 100410, 99530, 98720, 98360, 98350, 98350, 97970, 97870, 97770, 97740, 97250, 96980, 95550, 95510, 95060, 94790, 94610, 94390, 94240, 94020, 93370, 92500, 92460, 92360, 92070, 91370, 91190, 90970, 90640, 90420, 89850, 89280, 88890, 88510, 88510, 88410, 88270, 88040, 87910, 87420, 87380, 87340, 87100, 87090, 87020, 86850, 86800, 86140, 86090, 85880, 85880, 84770, 84300, 84060, 83200, 82770, 82510, 82450, 82240, 82090, 81690, 81630, 81580, 81430, 81390, 81230, 81100, 80970, 80570, 80440, 80370, 80300, 80120, 80050, 79990, 79970, 79710, 79650, 7950

In [30]:
#I would like to find out if the value of a job
#depends on the number of jobs available
!pip install plotly

distributed 1.21.8 requires msgpack, which is not installed.
You are using pip version 10.0.1, however version 18.0 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [98]:
#I create a chart to find out whether HR, Sales & Research and Analysis show a high demand in the market
#at the moment
#The chart below indicates that number of people employeed does not seem to have much  of an effect on salary hikes
#except for roles above USD140K. 
import plotly.plotly as py
plotly.offline.init_notebook_mode(connected=True)
# use offline mode to avoid initial registration

trace1 = dict(x=listSalary,
              y=listNumberEd, 
              mode="markers", 
              text=listJob,
             )

data = [trace1]

layout = dict(title = 'Top USA Jobs',
             xaxis = dict(title = 'Salary'),
             yaxis = dict(title = '# of Employed'),
                      )
#    marker = dict(color='#ffcdd2'),

fig = dict(data=data, layout=layout)


plotly.offline.iplot(fig)

#Human resources managers earn $110,120 and there are 136,310 jobs
#Sales managers earn $121,060 and there are 371,410 jobs
#There seems to be high salaries for doctors and medical practiotioners and not many of them in the market, so perhpaps this 
#could explain the reason for employees in the Research and development Department leaving.